lib/cli/test/unit/Cardano/CLISpec.hs

```diff
@@ -473,10 +558,132 @@ spec = do
+            \\"\175\\_(\12484)_/\175\". The full dictionary is available here:\
+            \ https://github.com/input-output-hk/cardano-wallet/tree/master/spe\
+            \cifications/mnemonic/english.txt\n")
+
+    describe "CliKeyScheme" $ do
+        it "all allowedWordLengths are supported"
+            $ property prop_allowedWordLengthsAllWork
+
+        it "scheme == scheme (reflexivity)" $ property $ \s ->
+            propCliKeySchemeEquality
+                (byronCliKeyScheme s)
+                (byronCliKeyScheme s)
+
+        -- This tests provides a stronger guarantee than merely knowing that
+        -- unsafeHexTextToXPrv and xPrvToHexText roundtrips.
+        it "scheme == mapKey (fromHex . toHex) scheme"
+            $ property prop_roundtripCliKeySchemeKeyViaHex
+
+        it "random /= icarus" $ do
+            expectFailure $ propCliKeySchemeEquality
+                (byronCliKeyScheme Random)
+                (byronCliKeyScheme Icarus)
+
   where
     backspace :: Text
     backspace = T.singleton (toEnum 127)
 
+prop_roundtripCliKeySchemeKeyViaHex :: ByronWalletStyle -> Property
+prop_roundtripCliKeySchemeKeyViaHex style =
+            propCliKeySchemeEquality
+                (byronCliKeyScheme style)
+                (mapKey unsafeHexTextToXPrv
+                    . mapKey xPrvToHexText
+                    $ byronCliKeyScheme style)
+
+prop_allowedWordLengthsAllWork :: ByronWalletStyle -> Property
+prop_allowedWordLengthsAllWork style = do
+    (forAll (genAllowedMnemonic s) propCanRetrieveRootKey)
+  where
+    s :: CliKeyScheme XPrv (Either String)
+    s = byronCliKeyScheme style
+
+    propCanRetrieveRootKey :: [Text] -> Property
+    propCanRetrieveRootKey mw = case mnemonicToRootKey s mw of
+        Right _ -> property True
+        Left e -> counterexample
+            (show (length mw) ++ " words, failed with: " ++ e)
+            (property False)
+
+propCliKeySchemeEquality
+    :: CliKeyScheme XPrv (Either String)
+    -> CliKeyScheme XPrv (Either String)
+    -> Property
+propCliKeySchemeEquality s1 s2 = do
+    (forAll (genAllowedMnemonic s1) propSameMnem)
+    .&&.
+    (allowedWordLengths s1) === (allowedWordLengths s2)
+  where
+    propSameMnem :: [Text] -> Property
+    propSameMnem mw = (mnemonicToRootKey s1 mw) === (mnemonicToRootKey s2 mw)
+
+genAllowedMnemonic :: CliKeyScheme key m -> Gen [Text]
+genAllowedMnemonic s = oneof (map genMnemonicOfSize $ allowedWordLengths s)
+
+genMnemonicOfSize :: Int -> Gen [Text]
+genMnemonicOfSize 12 = mnemonicToText <$> genMnemonic @12
+genMnemonicOfSize 15 = mnemonicToText <$> genMnemonic @15
+genMnemonicOfSize 18 = mnemonicToText <$> genMnemonic @18
+genMnemonicOfSize 21 = mnemonicToText <$> genMnemonic @21
+genMnemonicOfSize 24 = mnemonicToText <$> genMnemonic @24
+genMnemonicOfSize n = error $ "when this test was written, "
+    ++ show n ++ " was not a valid length of a mnemonic"
```

Comment:
style suggestion, use LambdaCase to avoid repeating the function name 6 times :grimacing: